Tests/SonarScanner.MSBuild.PreProcessor.Test/JreCaching/TarGzUnpackTests.cs

```diff
@@ -0,0 +1,126 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.Text;
+using FluentAssertions;
+using ICSharpCode.SharpZipLib.Core;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NSubstitute;
+using SonarScanner.MSBuild.Common;
+using SonarScanner.MSBuild.PreProcessor.JreCaching;
+using TestUtilities;
+
+namespace SonarScanner.MSBuild.PreProcessor.Test.JreCaching;
+
+[TestClass]
+public class TarGzUnpackTests
+{
+    private readonly IFileWrapper fileWrapper = Substitute.For<IFileWrapper>();
+    private readonly IDirectoryWrapper directoryWrapper = Substitute.For<IDirectoryWrapper>();
+    private readonly IOperatingSystemProvider osProvider = Substitute.For<IOperatingSystemProvider>();
+
+    [TestMethod]
+    public void TarGzUnpacking_Success()
+    {
+        // A tarball with the following content:
+        // Main
+        //  ├── Sub
+        //  └── Sub2
+        //      └── Sample.txt
+        const string sampleTarGzFile = """
+            H4sICL04jWYEAE1haW4udGFyAO3SUQrDIAyA4RzFE2wao55iTz2BBccK3Ribw
+            nb7iVDKnkqh+mK+l4S8/rn46XGGumTmnMuz+JvLrsiSRk1ImO/WkgRhoIH0jv
+            4lBHSq9B/SWPMHdvVHk+9OO+T+LSz9seID7OpPpT8Zy/1bWPsP/v6cwyl+Ihx
+            ss78ya3+T70qR1iAkNNB5/1v4ijH4FKdrmoExxlgvfmqGu7oADgAA
+            """;
+        var baseDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+        using var archive = new MemoryStream(Convert.FromBase64String(sampleTarGzFile));
+        using var unzipped = new MemoryStream();
+        fileWrapper.Create($"""{baseDirectory}\Main\Sub2\Sample.txt""").Returns(unzipped);
+
+        CreateUnpacker().Unpack(archive, baseDirectory);
+
+        directoryWrapper.Received(1).CreateDirectory($"""{baseDirectory}\Main\""");
+        directoryWrapper.Received(1).CreateDirectory($"""{baseDirectory}\Main\Sub\""");
+        directoryWrapper.Received(1).CreateDirectory($"""{baseDirectory}\Main\Sub2\""");
+        Encoding.UTF8.GetString(unzipped.ToArray()).NormalizeLineEndings().Should().Be("hey beautiful");
+    }
+
+    [TestMethod]
+    public void TarGzUnpacking_RootedPath_Success()
+    {
+        // A tarball with a single file with a rooted path: "\ sample.txt"
+        const string zipWithRootedPath = """
+            H4sIAAAAAAAAA+3OMQ7CMBBE0T3KngCtsY0PwDVoUlghkiEoNhLHB
+            5QmFdBEEdJ/zRQzxZy0dpdbybv2aLISe0kpvdOlaMucuSAuHIKPto
+            /eizmXfBS1tQ4t3WvrJlXpp9x/2n3r/9Q5lzLqcaxtuG79BQAAAAA
+            AAAAAAAAAAADwuyfh1ptHACgAAA==
+            """;
+        var baseDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+        using var unzipped = new MemoryStream();
+        fileWrapper.Create($"""{baseDirectory}\ sample.txt""").Returns(unzipped);
```

Comment:
Do we get calls to `directoryWrapper.Create` here? Add assertions for it, like e.g. `directoryWrapper.DidNotReceivedWithAnyArgs().Create(null);`

Reply:
We do get one call for the base directory. Added it.